09-schema/03-rules.md

```diff
@@ -274,6 +274,112 @@ TypeQLDefine query = TypeQL.define(
+when{
+    (from: $x, to: $y) isa edge;
+} then {
+    (from: $x, to: $y) isa forward-reachable;
+};
+
+rule forward-transitivity-recursive:
+when{
+    (from: $x, to: $y) isa forward-reachable;
+    (from: $y, to: $z) isa edge;
+} then {
+    (from: $x, to: $z) isa forward-reachable;
+};
+```
+
+We can intepret this as finding a path and extend it by one. Querying all nodes reachable from p in the chain p-q-r-s-t would generate the following relations:
+```
+p-q, q-r, r-s, s-t      (edges already in the database)
+
+p--q,                   (Inferred with the first rule)
+p--r,                   (Inferred with the second rule)
+p--s,                   (Inferred with the second rule)
+p--t                    (Inferred with the second rule)
+```
+Here, we only generate one relation for _**each node**_ reachable from p, bringing the complexity down from quadratic in to linear in the number of reachable nodes.
+
+### Backward transitivity
+To see what happens when we try to compute backwards transitivity using the above formulation, consider the query to find all nodes from which `t` is reachable in the same chain `p-q-r-s-t`. The second rule is now executed backwards - first checking all nodes `$y` from which there is an edge to `t`. Then it recursively queries all nodes reachable from `$y`. Thus, a relation is generated for every pair of nodes which are reachable from `t`.
+
+To answer backward transitive queries, we simply need a backwards version of the transitive relation and rules. Intuitively, This approach computes forward-transitivity on the reversed graph.
+```typeql
+define
+rule backward-transitivity-base:
+when{
+    (to: $x, from: $y) isa edge;
+} then {
+    (to: $x, from: $y) isa backward-reachable;
+};
+
+rule backward-transitivity-recursive:
+when{
+    (to: $x, from: $y) isa backward-reachable;
+    (to: $y, from: $z) isa edge;
+} then {
+    (to: $x, from: $z) isa backward-reachable;
+};
+```
+
+### Undirected transitivity
+We can use the same formulation for undirected graphs. If the undirected edges are defined by the relations `(node: $x, node: $y) isa edge;` then the rules would read:
+
+```typeql
+define
+rule forward-transitivity-base:
+when{
+    (node: $x, node: $y) isa edge;
+} then {
+    (from: $x, to: $y) isa forward-reachable;
+};
+
+rule forward-transitivity-recursive:
+when{
+    (from: $x, to: $y) isa forward-reachable;
+    (node: $y, node: $z) isa edge;
+} then {
+    (from: $x, to: $z) isa forward-reachable;
+};
+```
+ Although we still need different roles for `$x` and `$z` in the inferred relation, (without which we'd query the rule once in each direction) `forward-reachable` is actually equivalent to the undirected transitive relation.
+
+<div class = "note">
+[Important]
+* These rules are efficient only when evaluated with `$x` specified. Thus, when writing a query, it is recommended to use `forward-reachable` or `backward-reachable` depending on whether the `from` or `to` is specified.
```

Comment:
this is a key part of the situation, and should be embedded in the doc throughout + at the start. After doing that, we can delete this note.